services/sky/document_view.h

```diff
@@ -109,29 +92,31 @@ class DocumentView : public blink::ServiceProvider,
   void LoadFromSnapshotStream(String name,
                               mojo::ScopedDataPipeConsumerHandle snapshot);
 
-  void UpdateRootSizeAndViewportMetrics(const mojo::Rect& new_bounds);
+  void UpdateViewportMetrics(const mojo::ViewportMetrics& viewport_metrics);
+
+  void HandleInputEvent(mojo::EventPtr event);
 
   void InitServiceRegistry();
+  void InitViewport();
 
   mojo::URLResponsePtr response_;
   mojo::ServiceProviderImpl exported_services_;
   mojo::ServiceProviderPtr imported_services_;
-  mojo::InterfaceRequest<mojo::ServiceProvider> services_provided_to_embedder_;
-  mojo::ServiceProviderPtr services_provided_by_embedder_;
+  mojo::NativeViewportPtr viewport_service_;
   mojo::Shell* shell_;
   mojo::asset_bundle::AssetBundlePtr root_bundle_;
   mojo::NavigatorHostPtr navigator_host_;
   std::unique_ptr<blink::SkyView> sky_view_;
-  mojo::View* root_;
-  mojo::ViewManagerClientFactory view_manager_client_factory_;
+  scoped_ptr<DartLibraryProviderImpl> library_provider_;
   scoped_ptr<LayerHost> layer_host_;
   scoped_refptr<TextureLayer> root_layer_;
   std::unique_ptr<compositor::LayerTree> current_layer_tree_;  // TODO(abarth): Integrate //sky/compositor and //services/sky/compositor.
   compositor::PaintContext paint_context_;
   RasterizerBitmap* bitmap_rasterizer_;  // Used for pixel tests.
   mojo::ServiceRegistryPtr service_registry_;
-  scoped_ptr<mojo::StrongBinding<mojo::ServiceProvider>>
-      service_registry_service_provider_binding_;
+  mojo::Binding<NativeViewportEventDispatcher> event_dispatcher_binding_;
+  mojo::ViewportMetricsPtr viewport_metrics_;
```

Comment:
mojo::ViewportMetrics is a simple struct (data only).  I think you should store an instance by value here and assign from the passed in pointer rather than storing a pointer to the heap.  Then you won't have to Clone() copies - you can just use value semantics.